File: src/main/java/br/com/shoppinglist/shopping_list/identity/application/usecase/RequestRegistrationCodeUseCaseImpl.java
package br.com.shoppinglist.shopping_list.identity.application.usecase;

import br.com.shoppinglist.shopping_list.identity.application.dto.EmailVerificationCode;
import br.com.shoppinglist.shopping_list.identity.application.dto.MessageResult;
import br.com.shoppinglist.shopping_list.identity.application.dto.RequestRegistrationCodeCommand;
import br.com.shoppinglist.shopping_list.identity.application.port.in.RequestRegistrationCodeUseCase;
import br.com.shoppinglist.shopping_list.identity.application.port.out.EmailVerificationCodeRepository;
import br.com.shoppinglist.shopping_list.identity.application.port.out.PasswordHasher;
import br.com.shoppinglist.shopping_list.identity.application.port.out.RegistrationVerificationSettings;
import br.com.shoppinglist.shopping_list.identity.application.port.out.UserRepository;
import br.com.shoppinglist.shopping_list.identity.application.port.out.VerificationCodeSender;
import br.com.shoppinglist.shopping_list.identity.domain.exception.EmailAlreadyInUseException;
import java.security.SecureRandom;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import org.springframework.stereotype.Service;

@Service
public class RequestRegistrationCodeUseCaseImpl implements RequestRegistrationCodeUseCase {

	private static final SecureRandom RANDOM = new SecureRandom();

	private final UserRepository userRepository;
	private final EmailVerificationCodeRepository emailVerificationCodeRepository;
	private final PasswordHasher passwordHasher;
	private final VerificationCodeSender verificationCodeSender;
	private final RegistrationVerificationSettings verificationSettings;

	public RequestRegistrationCodeUseCaseImpl(
		UserRepository userRepository,
		EmailVerificationCodeRepository emailVerificationCodeRepository,
		PasswordHasher passwordHasher,
		VerificationCodeSender verificationCodeSender,
		RegistrationVerificationSettings verificationSettings
	) {
		this.userRepository = userRepository;
		this.emailVerificationCodeRepository = emailVerificationCodeRepository;
		this.passwordHasher = passwordHasher;
		this.verificationCodeSender = verificationCodeSender;
		this.verificationSettings = verificationSettings;
	}

	@Override
	public MessageResult execute(RequestRegistrationCodeCommand command) {
		String email = RegistrationInputValidator.normalizeEmail(command.email());
		RegistrationInputValidator.validateEmail(email);

		if (userRepository.existsByEmail(email)) {
			throw new EmailAlreadyInUseException(email);
		}

		String plainCode = generateSixDigitCode();
		long expirationMinutes = Math.max(verificationSettings.codeExpirationMinutes(), 1);
		Instant expiresAt = Instant.now().plus(expirationMinutes, ChronoUnit.MINUTES);

		EmailVerificationCode verificationCode = new EmailVerificationCode(
			email,
			passwordHasher.hash(plainCode),
			expiresAt
		);
		emailVerificationCodeRepository.save(verificationCode);
		verificationCodeSender.send(email, plainCode);

		return new MessageResult("Verification code sent to your email");
	}

	private String generateSixDigitCode() {
		int value = RANDOM.nextInt(1_000_000);
		return String.format("%06d", value);
	}
}

File: src/identity/application/usecase/request-registration-code.use-case.impl.ts
import { Injectable } from '@nestjs/common';
import { randomInt } from 'crypto';
import { EmailAlreadyInUseException } from '../../domain/exception/identity.exceptions';
import { MessageResult } from '../dto/message.result';
import { RequestRegistrationCodeCommand } from '../dto/request-registration-code.command';
import { RequestRegistrationCodeUseCase } from '../port/in/request-registration-code.use-case';
import { EmailVerificationCodeRepository } from '../port/out/email-verification-code.repository';
import { PasswordHasher } from '../port/out/password-hasher';
import { RegistrationVerificationSettings } from '../port/out/registration-verification.settings';
import { UserRepository } from '../port/out/user.repository';
import { VerificationCodeSender } from '../port/out/verification-code-sender';
import { RegistrationInputValidator } from './registration-input.validator';

@Injectable()
export class RequestRegistrationCodeUseCaseImpl extends RequestRegistrationCodeUseCase {
  constructor(
    private readonly users: UserRepository,
    private readonly verificationCodes: EmailVerificationCodeRepository,
    private readonly passwordHasher: PasswordHasher,
    private readonly verificationCodeSender: VerificationCodeSender,
    private readonly settings: RegistrationVerificationSettings,
  ) {
    super();
  }

  async execute(command: RequestRegistrationCodeCommand): Promise<MessageResult> {
    const email = RegistrationInputValidator.normalizeEmail(command.email);
    RegistrationInputValidator.validateEmail(email);

    if (await this.users.existsByEmail(email)) {
      throw new EmailAlreadyInUseException(email);
    }

    const plainCode = String(randomInt(0, 1_000_000)).padStart(6, '0');
    const expirationMinutes = Math.max(this.settings.codeExpirationMinutes(), 1);
    const expiresAt = new Date(Date.now() + expirationMinutes * 60_000);

    await this.verificationCodes.save({
      email,
      codeHash: await this.passwordHasher.hash(plainCode),
      expiresAt,
    });
    await this.verificationCodeSender.send(email, plainCode);

    return { message: 'Verification code sent to your email' };
  }
}
